File: backend/app/models/bulk_job.py
import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Enum, ForeignKey, Integer, String, TIMESTAMP, Text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database import Base


class BulkJobStatus(str, enum.Enum):
    pending   = "pending"    # processando plano de ação via IA
    ready     = "ready"      # plano gerado, aguardando execução
    executing = "executing"  # executando nos devices
    partial   = "partial"    # alguns ok, alguns falharam
    completed = "completed"  # todos concluídos
    failed    = "failed"     # todos falharam


class BulkJob(Base):
    __tablename__ = "bulk_jobs"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False, index=True,
    )
    created_by: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id", ondelete="RESTRICT"),
        nullable=False,
    )
    description: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[BulkJobStatus] = mapped_column(
        Enum(BulkJobStatus, native_enum=False), nullable=False, default=BulkJobStatus.pending
    )
    device_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    completed_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    failed_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    intent: Mapped[str | None] = mapped_column(String(100), nullable=True)
    error_summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

File: frontend/src/types/bulk_job.ts
import type { Operation } from "./operation";

export type BulkJobStatus =
  | "pending"
  | "ready"
  | "executing"
  | "partial"
  | "completed"
  | "failed";

export interface BulkJob {
  id: string;
  tenant_id: string;
  created_by: string;
  description: string;
  status: BulkJobStatus;
  device_count: number;
  completed_count: number;
  failed_count: number;
  intent: string | null;
  error_summary: string | null;
  created_at: string;
  updated_at: string;
}

export interface BulkJobDetail extends BulkJob {
  operations: Operation[];
}

export interface BulkJobCreate {
  device_ids: string[];
  natural_language_input: string;
}
